### PR TITLE
Cartesian grid changes

### PR DESCRIPTION
**Pull request by morozovkirill**
## Description
New props for CartesianGrid component:
syncWithTicks: boolean - if false, adds extra lines at min and max values (current default behaviour), if true - draws only lines at axis's ticks
horizontalValues:  number[] | string[] - custom values for horizontal grid lines
verticalValues?: number[] | string[] - custom values for vertical grid lines

## Related Issue
https://github.com/recharts/recharts/issues/2153

## Motivation and Context
More than 10 people voted for this feature in issues.

## How Has This Been Tested?
I've tested this in my pet project.

## Screenshots (if appropriate):
Снимок экрана 2023-09-08 в 19 25 25
Снимок экрана 2023-09-08 в 19 26 27

Снимок экрана 2023-09-08 в 19 19 49
Снимок экрана 2023-09-08 в 19 20 22

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: src/cartesian/CartesianGrid.tsx [hunk 1]
```diff
@@ -37,6 +37,9 @@ interface CartesianGridProps extends InternalCartesianGridProps {
   verticalPoints?: number[];
   verticalFill?: string[];
   horizontalFill?: string[];
+  syncWithTicks?: boolean;
+  horizontalValues?: number[] | string[];
+  verticalValues?: number[] | string[];
```
Comment (nikolasrieble) [inline src/cartesian/CartesianGrid.tsx:42]: This is a great PR, it adds value and is concise. Lets still reflect a second on the API. Is this consistent with the rest of the API? Can it be done more concisely?

Questions: 

(1) Why would the horizontalValues be a string? 
(2) What do `horizontalPoints` do vs `horizontalValues`? Both seem quite generic in name.
(3) Do you see syncWithTicks to be extended in the near future? What are feature that we already know we want to add here? (are there any that you could think of)

Change request: 
- Please add comments above each new field in the form of
```
  /** Some helpful explanation */
  syncWithTicks?: boolean;
```

These comments would then automatically show up in storybook in the ArgsTable, which would help us document the API.

**Review by nikolasrieble — CHANGES_REQUESTED** [review]
Please add comments to the type, and lets reflect on the API. It might be just fine, but once released it is hard to change, thus lets make sure we know the pros and cons of this design.

File: src/cartesian/CartesianGrid.tsx (hunk 1 above)
Comment (morozovkirill) [reply src/cartesian/CartesianGrid.tsx]: 1. YAxis type could be 'number' | 'category'. So, horizontal values could be strings, if we will have YAxis type as category.
2. HorizontalPoints are array of coordinates (in pixels), while ...values are axis values. I think it's hard sometimes to calc coordinates yourself.
3. Honestly, I have no idea about it.

File: src/cartesian/CartesianGrid.tsx [hunk 2]
```diff
@@ -37,8 +37,14 @@ interface CartesianGridProps extends InternalCartesianGridProps {
...
   horizontalFill?: string[];
+  /**
+   * If true, only the lines that correspond to the axes ticks values will be drawn.
+   * If false, extra lines could be added for each axis (at min and max coordinates), if there will not such ticks
+   */
   syncWithTicks?: boolean;
+  /** Array of values, where horizontal lines will be drawn. Numbers or strings, in dependence on axis type */
   horizontalValues?: number[] | string[];
+  /** Array of values, where vertical lines will be drawn. Numbers or strings, in dependence on axis type */
   verticalValues?: number[] | string[];
```
Comment (nikolasrieble) [inline src/cartesian/CartesianGrid.tsx:48]: How does this interact with 

```
  verticalPoints: {
    description: 'The x-coordinates of all vertical lines.',
    table: {
      type: {
        summary: 'Array',
        defaultValue: [],
      },
      category: 'General',
    },
  },
```

Comment (nikolasrieble) [reply src/cartesian/CartesianGrid.tsx:48]: What if both are passed? Which one takes precedence?

File: src/cartesian/CartesianGrid.tsx [hunk 3]
```diff
@@ -21,8 +21,8 @@ interface InternalCartesianGridProps {
   y?: number;
   width?: number;
   height?: number;
-  horizontalCoordinatesGenerator?: (props: any) => number[];
-  verticalCoordinatesGenerator?: (props: any) => number[];
+  horizontalCoordinatesGenerator?: (props: any, syncWithTicks: Boolean) => number[];
```
Comment (nikolasrieble) [inline src/cartesian/CartesianGrid.tsx:24]:
```suggestion
  horizontalCoordinatesGenerator?: (props: any, syncWithTicks: boolean) => number[];
```

File: src/cartesian/CartesianGrid.tsx [hunk 4]
```diff
@@ -21,8 +21,8 @@ interface InternalCartesianGridProps {
...
+  verticalCoordinatesGenerator?: (props: any, syncWithTicks: Boolean) => number[];
```
Comment (nikolasrieble) [inline src/cartesian/CartesianGrid.tsx:25]:
```suggestion
  verticalCoordinatesGenerator?: (props: any, syncWithTicks: boolean) => number[];
```

Comment (nikolasrieble) [reply src/cartesian/CartesianGrid.tsx:25]: https://stackoverflow.com/questions/64443288/what-is-the-difference-between-boolean-and-boolean-in-typescript

File: src/util/ChartUtils.ts [hunk 5]
```diff
@@ -537,12 +537,17 @@ export const isCategoricalAxis = (layout: LayoutType | PolarLayoutType, axisType
...
 /**
  * Calculate the Coordinates of grid
- * @param  {Array} ticks The ticks in axis
- * @param {Number} min   The minimun value of axis
- * @param {Number} max   The maximun value of axis
- * @return {Array}       Coordinates
+ * @param  {Array} ticks           The ticks in axis
+ * @param {Number} min             The minimun value of axis
+ * @param {Number} max             The maximun value of axis
+ * @param {Boolean} syncWithTicks  Synchronize grid lines with ticks or not
```
Comment (nikolasrieble) [inline src/util/ChartUtils.ts:543]:
```suggestion
 * @param {boolean} syncWithTicks  Synchronize grid lines with ticks or not
```

File: src/cartesian/CartesianGrid.tsx (hunk 2 above)
Comment (nikolasrieble) [reply src/cartesian/CartesianGrid.tsx:48]: Not required for this PR, but it would be great if you could add comments to the other props as well, and ideally clarify the question above, i.e. declare precedence in the comment, because that would directly land in the documentation. 

Not at all required, I am just thinking out loud.

File: src/cartesian/CartesianGrid.tsx (hunk 3 above)
Comment (ckifer) [reply src/cartesian/CartesianGrid.tsx:24]: +1

File: src/cartesian/CartesianGrid.tsx (hunk 4 above)
Comment (ckifer) [reply src/cartesian/CartesianGrid.tsx:25]: +1

File: src/cartesian/CartesianGrid.tsx (hunk 2 above)
Comment (ckifer) [reply src/cartesian/CartesianGrid.tsx:48]: Yeah I'm actually confused about this. Are not `verticalPoints`/`horizontalPoints` the way that the grid internally generates its lines?

Comment (morozovkirill) [reply src/cartesian/CartesianGrid.tsx:63]: The priority is: if we have vertical/horizontal points, we use them; then vertical/horizontal values follow; and finally, syncWithTicks, if none of points/values passed.

I added comments about this in CartesianGrid.tsx

Comment (nikolasrieble) [reply src/cartesian/CartesianGrid.tsx:63]: So, should the type possibly be: 

```
type Config = {
 verticalPoints ..
 horizontalPoints ..
} | {
 verticalValues ..
 horizontalValues ..
} | {
 syncWithTicks ..
}
```

Would that make sense? It could nicely communicate our TS users that using them together does not make sense.

Comment (morozovkirill) [reply src/cartesian/CartesianGrid.tsx:63]: @nikolasrieble I think you are right, but I don't know how it could be implemented. I think, my knowledge is not enough.

File: package-lock.json [hunk 6]
```diff
@@ -26439,6 +26439,7 @@
       "version": "16.14.0",
```
Comment (nikolasrieble) [inline package-lock.json:26439]: How come there are changes in the package-lock.json? 🤔 These do seem related.

File: src/cartesian/CartesianGrid.tsx (hunk 2 above)
Comment (nikolasrieble) [reply src/cartesian/CartesianGrid.tsx:63]: I believe it would be too much to ask from you to fix our typing, especially given that the current typing at Recharts is okay at best. 

I still wanted to share this as the direction I would have preferred.

**Review by nikolasrieble — COMMENTED** [review]
@morozovkirill Thank you for this great contribution. A new feature to further make recharts more customisable, a step consistent with the rest of the plattform. Added documentation, just perfect. 

Also thank you for meeting all our comments and questions. The only missing thing is the unrelated changes in the package-lock.json. Please revert them, and then lets land this 💪 ❤️ 

EDIT: 
The results from a bit more testing seem to show that this does not work as intended. Please make sure to adapt the storybook story such that this can be showcased. See example below.

**Comment by nikolasrieble** [comment]
Testing this in storybook. 

I adapted the storybook story for the CartesianGrid as such: 

```
--- a/storybook/stories/API/cartesian/CartesianGrid.stories.tsx
+++ b/storybook/stories/API/cartesian/CartesianGrid.stories.tsx
@@ -1,6 +1,7 @@
 import React from 'react';
 import { Args } from '@storybook/react';
-import { CartesianGrid, ResponsiveContainer, ComposedChart } from '../../../../src';
+import { CartesianGrid, ResponsiveContainer, ComposedChart, XAxis, YAxis, Line } from '../../../../src';
+import { pageData } from '../../data';
 
 const GeneralProps: Args = {
   x: {
@@ -107,22 +108,25 @@ export const API = {
     const [surfaceWidth, surfaceHeight] = [500, 500];
     return (
       <ResponsiveContainer width="100%" height={surfaceHeight}>
-        <ComposedChart width={surfaceWidth} height={surfaceHeight}>
+        <ComposedChart data={pageData}>
+          <XAxis dataKey={'name'} />
+          <YAxis />
+          <Line dataKey={'uv'} />Í
           <CartesianGrid {...args} />
         </ComposedChart>
       </ResponsiveContainer>
     );
   },
   args: {

```

To better show the feature. A few findings: 

(1) There seems to be no difference between syncWithTicks and the default behaviour.
(2) HorizontalValues seems to work. When it set it to an array including numeric values, it draws the ticks exactly at the point where the values are (numeric axis on YAxis) 
(3) VerticalValues does not seem to work with a categorical axis, such as in the default toy example for recharts.

File: package-lock.json (hunk 6 above)
Comment (morozovkirill) [reply package-lock.json]: I've fixed it

**Comment by nikolasrieble** [comment]
@morozovkirill This is a great contribution, almost done as it seems. Is there anything I could help with?

**Comment by morozovkirill** [comment]
> @morozovkirill This is a great contribution, almost done as it seems. Is there anything I could help with?

I don’t know really. What I should do next?

**Comment by nikolasrieble** [comment]
@morozovkirill During initial testing I could not find a difference between syncWithTicks and the default behaviour. When reading through the issue again I found that it is only an edge case for when this is really applicable. With the edge case it works as expected. 

Thus this is good to go. Sorry for the confusion. I missed context as to how to test this feature relevantly.

**Review by nikolasrieble — APPROVED** [review]
Thank you for your contribution! :heart:

File: src/cartesian/CartesianGrid.tsx [hunk 7]
```diff
@@ -276,12 +303,38 @@ export class CartesianGrid extends PureComponent<Props> {
 
     // No horizontal points are specified
     if ((!horizontalPoints || !horizontalPoints.length) && _.isFunction(horizontalCoordinatesGenerator)) {
-      horizontalPoints = horizontalCoordinatesGenerator({ yAxis, width: chartWidth, height: chartHeight, offset });
+      const isHorizontalValues = horizontalValues && horizontalValues.length;
+
+      horizontalPoints = horizontalCoordinatesGenerator(
+        {
+          yAxis: {
+            ...yAxis,
```
Comment (ckifer) [inline src/cartesian/CartesianGrid.tsx:311]: this needs an existence check

File: src/cartesian/CartesianGrid.tsx [hunk 8]
```diff
@@ -276,12 +303,38 @@ export class CartesianGrid extends PureComponent<Props> {
...
 
     // No vertical points are specified
     if ((!verticalPoints || !verticalPoints.length) && _.isFunction(verticalCoordinatesGenerator)) {
-      verticalPoints = verticalCoordinatesGenerator({ xAxis, width: chartWidth, height: chartHeight, offset });
+      const isVerticalValues = verticalValues && verticalValues.length;
+
+      verticalPoints = verticalCoordinatesGenerator(
+        {
+          xAxis: {
+            ...xAxis,
```
Comment (ckifer) [inline src/cartesian/CartesianGrid.tsx:329]: this needs an existence check
